gradio/routes.py

```diff
@@ -409,21 +411,22 @@ async def run_predict(
                         if getattr(block, "stateful", False)
                     }
                 session_state = app.state_holder[body.session_hash]
-                iterators = app.iterators[body.session_hash]
                 # The should_reset set keeps track of the fn_indices
                 # that have been cancelled. When a job is cancelled,
                 # the /reset route will mark the jobs as having been reset.
                 # That way if the cancel job finishes BEFORE the job being cancelled
                 # the job being cancelled will not overwrite the state of the iterator.
-                # In all cases, should_reset will be the empty set the next time
-                # the fn_index is run.
-                app.iterators[body.session_hash]["should_reset"] = set()
+                if fn_index in app.iterators_to_reset[body.session_hash]:
```

Comment:
Don't quite understand why you were having issues with cancelling! Regardless, event cancelling is working well for me in this branch as well so I think we're ok to merge this change in.